src/Illuminate/Database/Eloquent/Builder.php

```diff
@@ -896,21 +906,95 @@ public function applyScopes()
         }
 
         $builder = clone $this;
+        $query = $builder->getQuery();
+
+        $count = count($query->wheres);
+        $offsets = [0, $count];
 
         foreach ($this->scopes as $scope) {
-            if ($scope instanceof Closure) {
-                $scope($builder);
-            }
+            $this->applyScope($scope, $builder);
 
-            if ($scope instanceof ScopeInterface) {
-                $scope->apply($builder, $this->getModel());
-            }
+            $offsets[] = count($query->wheres);
+        }
+
+        if ($this->shouldNestWheresForScope($count, $query)) {
+            $this->nestWheresForScope($query, array_unique($offsets));
         }
 
         return $builder;
     }
 
     /**
+     * Apply a single scope on the given builder instance.
+     *
+     * @param  \Illuminate\Database\Eloquent\ScopeInterface|\Closure  $scope
+     * @param  \Illuminate\Database\Eloquent\Builder  $builder
+     * @return void
+     */
+    protected function applyScope($scope, $builder)
+    {
+        if ($scope instanceof Closure) {
+            $scope($builder);
+        }
+
+        if ($scope instanceof ScopeInterface) {
```

Comment:
Does matter really. Something could extend a closure and implement the scope interface.